### PR TITLE
protoc-gen-swagger: support JSON Schema Validation properties and add openapiv2_field option

### PR DESCRIPTION
**Pull request by co3k**
This change supports JSON Schema Validation properties.
It allows you to specify some validation rules in `.proto` file like this:

```proto
message GeoCoordinate {
  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
    json_schema: {
      required: ["latitude", "longitude"]
    }
  };

  string latitude = 1;
  string longitude = 2;
}
```

And also this supports openapiv2_field field option.
It allows you to specifiy additional JSON schema setting like this:

```proto
string latitude = 1 [(grpc.gateway.protoc_gen_swagger.options.openapiv2_field) = {pattern: "^[\-]?[0-9]{0,2}\.[0-9]+$"}];
string longitude = 2 [(grpc.gateway.protoc_gen_swagger.options.openapiv2_field) = {pattern: "^[\-]?[0-9]{0,3}\.[0-9]+$"}];
```

The both of `grpc.gateway.protoc_gen_swagger.options.openapiv2_schema` and `grpc.gateway.protoc_gen_swagger.options.openapiv2_field` now supports the following validation properties.

* multiple_of
* maximum
* exclusive_maximum
* minimum
* exclusive_minimum
* max_length
* min_length
* pattern
* max_items
* min_items
* unique_items
* max_properties
* min_properties
* required

Please see also [JSON Schema Validation spec](https://tools.ietf.org/html/draft-fge-json-schema-validation-00) to understand about these properties.

### CODE REVIEW

**Comment by co3k** [comment]
Hi, I really need these changes for my works but there isn't any reactions for a month.
What should I do to get any feedbacks about this pull request?

**Comment by johanbrandhorst** [comment]
This is really interesting, but to make things easier for us maintainers, could you please include a description of the changes? Include any external links as well.

**Comment by co3k** [comment]
@johanbrandhorst Thanks for your advice! I've added descriptions to this pull request. If it's OK, I will also change the commit message to follow it.

**Comment by johanbrandhorst** [comment]
@co3k that's great thanks! I think this looks good but I would like the input of @ivucica as well before we merge.

File: protoc-gen-swagger/options/annotations.proto [hunk 1]
```diff
@@ -35,3 +35,10 @@ extend google.protobuf.ServiceOptions {
   // different descriptor messages.
   Tag openapiv2_tag = 1042;
 }
+extend google.protobuf.FieldOptions {
+  // ID assigned by protobuf-global-extension-registry@google.com for grpc-gateway project.
+  //
+  // All IDs are the same, as assigned. It is okay that they are the same, as they extend
+  // different descriptor messages.
+  JSONSchema openapiv2_field_json_schema = 1042;
```
Comment (ivucica) [inline protoc-gen-swagger/options/annotations.proto]: Can we name this just `openapiv2_field` similar to what was done with other options?

Comment (co3k) [reply protoc-gen-swagger/options/annotations.proto]: I've fixed it in https://github.com/grpc-ecosystem/grpc-gateway/pull/687/commits/48e32c98afbc0758fbb4c635e0f67b601532f49c

**Comment by co3k** [comment]
I wonder why the `Go: 1.10.x` `GATEWAY_PLUGIN_FLAGS=` Travis CI build gets failed. After my rebase. I've tried `make realclean examples` but this problem isn't solved.

**Comment by johanbrandhorst** [comment]
@co3k There's definitely something weird up with our CI pipeline at the moment, I'm trying to work out with the other maintainers what we need to do, sorry for the inconvenience. We'll get this change in as soon as possible!

**Comment by co3k** [comment]
Oh, I'm sorry...I've forgotten to do `go get -u github.com/golang/protobuf/protoc-gen-go` before my local building. 

Finally, Travis CI builds have been passed.

**Comment by johanbrandhorst** [comment]
> `go get -u github.com/golang/protobuf/protoc-gen-go`

You absolutely should not be expected to do this. Install it from the vendor folder: `go install ./vendor/github.com/golang/protobuf/protoc-gen-go`

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Comment by achew22** [comment]
@johanbrandhorst would it be possible to have the `Makefile` build `protoc-gen-go` from the vendor directory and use it instead?

**Comment by johanbrandhorst** [comment]
@achew22 This was part of the change I made already: https://github.com/grpc-ecosystem/grpc-gateway/blob/221ef342721da51038fa6fc3fbd4f0d190f9f6f7/Makefile#L119

**Pull request merged**